scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

```diff
@@ -0,0 +1,527 @@
+package com.twitter.scalding.typed.memory_backend
+
+import cascading.flow.{ FlowDef, FlowConnector }
+import cascading.pipe.Pipe
+import cascading.tap.Tap
+import cascading.tuple.{ Fields, TupleEntryIterator }
+import com.twitter.scalding.typed._
+import com.twitter.scalding.{ Config, Execution, ExecutionCounters, Mode }
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ ArrayList, Collections, UUID }
+import scala.concurrent.{ Future, ExecutionContext => ConcurrentExecutionContext, Promise }
+import scala.collection.mutable.{ ArrayBuffer, Map => MMap }
+import scala.collection.JavaConverters._
+
+import Execution.{ ToWrite, Writer }
+
+class AtomicBox[T <: AnyRef](init: T) {
+  private[this] val ref = new AtomicReference[T](init)
+
+  def lazySet(t: T): Unit =
+    ref.lazySet(t)
+
+  /**
+   * use a pure function to update the state.
+   * fn may be called more than once
+   */
+  def update[R](fn: T => (T, R)): R = {
+
+    @annotation.tailrec
+    def loop(): R = {
+      val init = ref.get
+      val (next, res) = fn(init)
+      if (ref.compareAndSet(init, next)) res
+      else loop()
+    }
+
+    loop()
+  }
+
+  def get(): T = ref.get
+}
+
+final class MemoryMode private (srcs: Map[TypedSource[_], Iterable[_]], sinks: Map[TypedSink[_], AtomicBox[Option[Iterable[_]]]]) extends Mode {
+
+  def newWriter(): Writer =
+    new MemoryWriter(this)
+
+  def openForRead(config: Config, tap: Tap[_, _, _]): TupleEntryIterator = ???
+  def fileExists(filename: String): Boolean = ???
+  def newFlowConnector(props: Config): FlowConnector = ???
+
+  def addSource[T](src: TypedSource[T], ts: Iterable[T]): MemoryMode =
+    new MemoryMode(srcs + (src -> ts), sinks)
+
+  def addSink[T](sink: TypedSink[T]): MemoryMode =
+    new MemoryMode(srcs, sinks + (sink -> new AtomicBox[Option[Iterable[_]]](None)))
+
+  /**
+   * This has a side effect of mutating this MemoryMode
+   */
+  def writeSink[T](t: TypedSink[T], iter: Iterable[T]): Unit =
+    sinks(t).lazySet(Some(iter))
+
+  def readSink[T](t: TypedSink[T]): Option[Iterable[T]] =
+    sinks.get(t).flatMap(_.get).asInstanceOf[Option[Iterable[T]]]
+
+  def readSource[T](t: TypedSource[T]): Option[Iterable[T]] =
+    srcs.get(t).asInstanceOf[Option[Iterable[T]]]
+}
+
+object MemoryMode {
+  def empty: MemoryMode = new MemoryMode(Map.empty, Map.empty)
+}
+
+object MemoryPlanner {
+
+  sealed trait Op[+O] {
+    def result(implicit cec: ConcurrentExecutionContext): Future[ArrayBuffer[_ <: O]]
+
+    def flatMap[O1](fn: O => TraversableOnce[O1]): Op[O1] =
```

Comment:
I hear you on `flatMap`. I can rename concatMap, but the fact is we used `flatMap` for scalding to make it a bit easier for newbies to scala (since it is like flatMap on List in that List accepts a similarly broad return type). `TypedPipe[T]` is definitely not a monad (it is Applicative!). But I can change this name here.